File: app/src/main/java/com/tablehockey/game/game/Skater.kt
package com.tablehockey.game.game

import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.sin

enum class Role(val label: String) { C("C"), LW("LW"), RW("RW"), LD("LD"), RD("RD"), G("G") }

/**
 * One player on the ice - five skaters and a goalie per team. Pure state;
 * all behaviour lives in [Simulation] / [AIController] and drawing in [Renderer].
 */
class Skater(val team: Int, val index: Int, val role: Role, val number: Int) {
    val isGoalie: Boolean get() = role == Role.G

    var x = 0f
    var y = 0f
    var vx = 0f
    var vy = 0f
    /** Facing direction in radians (world frame). */
    var facing = 0f

    val radius: Float = if (role == Role.G) 2.1f else 1.5f

    /** > 0 while knocked down after a body check. */
    var stunTimer = 0f
    /** > 0 while the stick is extended for a poke check. */
    var pokeTimer = 0f
    /** > 0 while lunging into a body check. */
    var checkTimer = 0f
    /** Generic cooldown between poke/hit attempts. */
    var actionCooldown = 0f
    /** Can't re-collect the puck until this expires (after shooting / passing). */
    var pickupCooldown = 0f
    /** Shot / pass animation. */
    var swingTimer = 0f
    /** Advances with distance skated; drives the stride animation. */
    var stride = 0f

    // AI scratch state
    var aiTimer = 0f
    var aiTargetX = 0f
    var aiTargetY = 0f
    var aiChaser = false

    // Goalie-only
    var butterfly = false
    /** Multiplier on the goalie's blocking area (difficulty). */
    var padScale = 1f

    // Network interpolation targets (client only)
    var netX = 0f
    var netY = 0f
    var netFacing = 0f

    val speed: Float get() = hypot(vx, vy)

    val stickReach: Float get() = radius + if (pokeTimer > 0f) POKE_REACH else STICK_REACH

    fun bladeX(): Float = x + cos(facing) * stickReach
    fun bladeY(): Float = y + sin(facing) * stickReach

    fun distanceTo(px: Float, py: Float): Float = hypot(px - x, py - y)

    fun place(px: Float, py: Float, face: Float) {
        x = px; y = py; vx = 0f; vy = 0f
        facing = face
        stunTimer = 0f; pokeTimer = 0f; checkTimer = 0f
        actionCooldown = 0f; pickupCooldown = 0f; swingTimer = 0f
        aiTimer = 0f; aiTargetX = px; aiTargetY = py; aiChaser = false
        butterfly = false
        netX = px; netY = py; netFacing = face
    }

    companion object {
        const val STICK_REACH = 2.6f
        const val POKE_REACH = 4.6f
        const val MAX_SPEED = 25f      // ft/s
        const val GOALIE_SPEED = 20f
    }
}

File: ios/PowerPlayHockeyCore/Sources/PowerPlayHockeyCore/Game/Skater.swift
import Foundation

enum Role {
    case c, lw, rw, ld, rd, g

    var label: String {
        switch self {
        case .c: return "C"
        case .lw: return "LW"
        case .rw: return "RW"
        case .ld: return "LD"
        case .rd: return "RD"
        case .g: return "G"
        }
    }
}

enum GoalieAction {
    case none
    case butterfly
    case padStack
}

/// One player on the ice - five skaters and a goalie per team. Pure state;
/// all behaviour lives in Simulation / AIController and drawing in Renderer.
final class Skater {
    static let STICK_REACH: Float = 2.6
    static let POKE_REACH: Float = 4.6
    static let MAX_SPEED: Float = 25      // ft/s
    static let GOALIE_SPEED: Float = 20

    let team: Int
    let index: Int
    let role: Role
    let number: Int

    var isGoalie: Bool { role == .g }

    var x: Float = 0
    var y: Float = 0
    var vx: Float = 0
    var vy: Float = 0
    /// Facing direction in radians (world frame).
    var facing: Float = 0

    let radius: Float

    /// > 0 while knocked down after a body check.
    var stunTimer: Float = 0
    /// > 0 while the stick is extended for a poke check.
    var pokeTimer: Float = 0
    /// > 0 while lunging into a body check.
    var checkTimer: Float = 0
    /// Generic cooldown between poke/hit attempts.
    var actionCooldown: Float = 0
    /// > 0 while performing a deke (lateral dodge/juke).
    var dekeTimer: Float = 0
    var dekeDir: Float = 1
    /// Can't re-collect the puck until this expires (after shooting / passing).
    var pickupCooldown: Float = 0
    /// Shot / pass animation.
    var swingTimer: Float = 0
    /// Advances with distance skated; drives the stride animation.
    var stride: Float = 0

    // AI scratch state
    var aiTimer: Float = 0
    var aiTargetX: Float = 0
    var aiTargetY: Float = 0
    var aiChaser: Bool = false

    // Goalie active save moves
    var butterfly: Bool = false
    var goalieAction: GoalieAction = .none
    var goalieActionTimer: Float = 0
    var padStackDir: Float = 1
    /// Multiplier on the goalie's blocking area (difficulty).
    var padScale: Float = 1

    // One-timer & penalty state
    var oneTimerArmed: Bool = false
    var inPenaltyBox: Bool = false
    var breathTimer: Float = 0

    // Network interpolation targets (client only)
    var netX: Float = 0
    var netY: Float = 0
    var netFacing: Float = 0

    init(team: Int, index: Int, role: Role, number: Int) {
        self.team = team
        self.index = index
        self.role = role
        self.number = number
        self.radius = role == .g ? 2.1 : 1.5
    }

    var speed: Float { hypot(vx, vy) }

    var stickReach: Float { radius + (pokeTimer > 0 ? Skater.POKE_REACH : Skater.STICK_REACH) }

    func bladeX() -> Float { x + cos(facing) * stickReach }
    func bladeY() -> Float { y + sin(facing) * stickReach }

    func distanceTo(_ px: Float, _ py: Float) -> Float { hypot(px - x, py - y) }

    func place(_ px: Float, _ py: Float, _ face: Float) {
        x = px; y = py; vx = 0; vy = 0
        facing = face
        stunTimer = 0; pokeTimer = 0; checkTimer = 0; dekeTimer = 0
        actionCooldown = 0; pickupCooldown = 0; swingTimer = 0
        aiTimer = 0; aiTargetX = px; aiTargetY = py; aiChaser = false
        butterfly = false; goalieAction = .none; goalieActionTimer = 0
        netX = px; netY = py; netFacing = face
    }
}
